File: src/understudy/store.py
"""In-memory event store with the derived indices the deterministic summary needs."""

from __future__ import annotations

from collections import Counter
from pathlib import Path

from understudy.events import Event, Kind


class EventStore:
    def __init__(self) -> None:
        self.events: list[Event] = []
        self.files_touched: Counter[str] = Counter()
        self.tool_counts: Counter[str] = Counter()
        self.turn_tool_counts: Counter[str] = Counter()
        self.last_tool: tuple[str, bool] | None = None
        self.last_action: str = "waiting…"
        self.current_turn: str | None = None
        self.error_count: int = 0

    def add(self, ev: Event) -> None:
        self.events.append(ev)
        payload = ev.payload
        if ev.kind == Kind.USER_PROMPT:
            self.current_turn = ev.turn_id
            self.turn_tool_counts = Counter()
            self.last_action = "reading user prompt"
        elif ev.kind == Kind.TOOL_CALL:
            name = payload.get("name", "tool")
            self.tool_counts[name] += 1
            self.turn_tool_counts[name] += 1
            self.last_action = f"calling {name}"
        elif ev.kind == Kind.TOOL_RESULT:
            ok = bool(payload.get("ok", True))
            self.last_tool = (payload.get("name", "tool"), ok)
            if not ok:
                self.error_count += 1
        elif ev.kind == Kind.FILE_EDIT:
            self.files_touched[payload.get("path", "")] += 1
            self.last_action = f"editing {Path(payload.get('path', '')).name}"
        elif ev.kind == Kind.THINKING:
            self.last_action = "thinking"
        elif ev.kind == Kind.ASSISTANT_TEXT:
            self.last_action = "responding"

    def bulk_add(self, events: list[Event]) -> None:
        for ev in events:
            self.add(ev)

File: crates/core/src/store.rs
//! In-memory event store with the derived indices the deterministic summary needs.

use std::collections::HashMap;

use crate::events::{Event, EventKind};

pub struct EventStore {
    pub events: Vec<Event>,
    pub files_touched: HashMap<String, usize>,
    pub tool_counts: HashMap<String, usize>,
    pub turn_tool_counts: HashMap<String, usize>,
    pub last_tool: Option<(String, bool)>,
    pub last_action: String,
    pub current_turn: Option<String>,
    pub error_count: usize,
}

impl Default for EventStore {
    fn default() -> Self {
        EventStore {
            events: Vec::new(),
            files_touched: HashMap::new(),
            tool_counts: HashMap::new(),
            turn_tool_counts: HashMap::new(),
            last_tool: None,
            last_action: "waiting…".to_string(),
            current_turn: None,
            error_count: 0,
        }
    }
}

impl EventStore {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add(&mut self, ev: Event) {
        match &ev.kind {
            EventKind::UserPrompt { .. } => {
                self.current_turn = ev.turn_id.clone();
                self.turn_tool_counts.clear();
                self.last_action = "reading user prompt".to_string();
            }
            EventKind::ToolCall { name, .. } => {
                *self.tool_counts.entry(name.clone()).or_default() += 1;
                *self.turn_tool_counts.entry(name.clone()).or_default() += 1;
                self.last_action = format!("calling {name}");
            }
            EventKind::ToolResult { name, ok, .. } => {
                self.last_tool = Some((name.clone(), *ok));
                if !*ok {
                    self.error_count += 1;
                }
            }
            EventKind::FileEdit { path, .. } => {
                *self.files_touched.entry(path.clone()).or_default() += 1;
                self.last_action = format!("editing {}", basename(path));
            }
            EventKind::Thinking { .. } => self.last_action = "thinking".to_string(),
            EventKind::AssistantText { .. } => self.last_action = "responding".to_string(),
            _ => {}
        }
        self.events.push(ev);
    }

    pub fn bulk_add(&mut self, events: Vec<Event>) {
        for ev in events {
            self.add(ev);
        }
    }
}

pub(crate) fn basename(path: &str) -> &str {
    path.rsplit(['/', '\\']).next().unwrap_or(path)
}
